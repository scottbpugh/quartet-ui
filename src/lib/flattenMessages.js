// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

export function flattenMessages(nestedMessages, prefix = "") {
  return Object.keys(nestedMessages).reduce((messages, key) => {
    const value = nestedMessages[key];
    const prefixedKey = prefix ? `${prefix}.${key}` : key;

    if (typeof value === "string") {
      messages[prefixedKey] = value;
    } else {
      Object.assign(messages, flattenMessages(value, prefixedKey));
    }

    return messages;
  }, {});
}

window.qu4rtet.exports("lib/flattenMessages", this);
