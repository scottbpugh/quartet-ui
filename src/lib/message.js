// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {Position, Toaster, Intent} from "@blueprintjs/core";
import {pluginRegistry} from "plugins/pluginRegistration";
import stringHash from "string-hash";

const MINIMUM_EXPIRE = 7000;

const msgToaster = Toaster.create({
  className: "my-toaster",
  position: Position.BOTTOM_RIGHT
});

const getIntent = type => {
  switch (type) {
    case "success":
      return Intent.SUCCESS;
    case "danger":
    case "error":
      return Intent.DANGER;
    case "warning":
      return Intent.WARNING;
    default:
      return Intent.PRIMARY;
  }
};

const _msgHashMap = {};

const showMessageNoRepeat = (toastMessage, expires_in = MINIMUM_EXPIRE) => {
  let msgHash = stringHash(JSON.stringify(toastMessage));
  let timestamp = new Date().getTime();
  if (msgHash in _msgHashMap && _msgHashMap[msgHash].expires > timestamp) {
    // don't show msg and update the expiration to prevent the same message
    // over and over.
    _msgHashMap[msgHash].expires = timestamp + expires_in;
    return;
  }
  _msgHashMap[msgHash] = {expires: timestamp + expires_in};
  msgToaster.show(toastMessage);
};

export const showMessage = msg => {
  try {
    // for errors etc out of a component context.
    const intl = pluginRegistry.getIntl();
    if (msg.id) {
      showMessageNoRepeat(
        {
          message: intl.formatMessage({id: msg.id}, msg.values),
          intent: getIntent(msg.type)
        },
        msg.expires_in || MINIMUM_EXPIRE
      );
    } else if (msg.msg) {
      //let formatMsg = typeof msg.msg === "object" ? JSON.stringify(msg.msg) : msg.msg;
      showMessageNoRepeat(
        {message: msg.msg, intent: getIntent(msg.type)},
        msg.expires_in || MINIMUM_EXPIRE
      );
    }
  } catch (e) {
    console.log(
      "an error occurred while attempting to display an error toast.",
      e
    );
  }
};

window.qu4rtet.exports("lib/message", this);
