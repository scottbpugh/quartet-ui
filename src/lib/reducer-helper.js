// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// A helper function meant to make life a little easier in reducers.

/**
 * Unknown - Description
 *
 * @param {object} [state={}]       The reducer state
 * @param {type}   serverID         The serverID
 * @param {object} [mergeObject={}] The object to merge into the state for this server.
 *
 * @return {type} Description
 */
export const setServerState = (state = {}, serverID, mergeObject = {}) => {
  if (!serverID) {
    throw new Error(
      "A server ID must be defined to set the server state in reducer."
    );
  }
  const copy = {...state}; // use a copy of state for manipulations.
  if (!copy.servers) {
    // if servers are not created yet, create an object.
    copy.servers = {};
  }
  if (!copy.servers[serverID]) {
    // if there is no object for this server, create it.
    copy.servers[serverID] = {};
  }
  return {
    ...copy,
    servers: {
      ...copy.servers,
      [serverID]: {
        ...copy.servers[serverID],
        ...mergeObject
      }
    }
  };
};
window.qu4rtet.exports("lib/reducer-helper", this);
