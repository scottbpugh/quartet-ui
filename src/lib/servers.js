// Copyright (c) 2018 Serial Lab
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import Swagger from "swagger-client";
import {pluginRegistry} from "plugins/pluginRegistration";
import {store} from "store";
import actions from "actions/serversettings";

/**
 * Server - Holds data about a server and its settings, API client, ...
 */
export class Server {
  constructor(serverSettings) {
    /*
    Following object should be passed.
    {
      serverSettingName,
      hostname,
      port,
      ssl,
      serverID,
      path,
      username,
      password
    }
    */
    this.setServerData(serverSettings);
    // make saved server object available to core and plugins.
    pluginRegistry.registerServer(this);
  }

  setServerData = serverSettings => {
    /*
    Following object should be passed.
    {
      serverSettingName,
      hostname,
      port,
      ssl,
      serverID,
      path,
      username,
      password
    */
    // allow for partial updates.
    for (let key in serverSettings) {
      if (serverSettings[key] !== "" || serverSettings[key] !== undefined) {
        this[key] = serverSettings[key];
      }
    }
    this.protocol = this.ssl === true ? "https" : "http";
    this.port = !this.port ? 80 : this.port;
    this.path = !this.path && !serverSettings.path ? "" : this.path;
    this.url = this.getServerURL();
    this._client = null;
    this.manifest = null;
    this.appList = [];
  };
  getFormStructure = () => {
    return Server.getFormStructure(this.toJSON());
  };
  static getFormStructure = (initialValues = {}) => [
    {
      name: "serverSettingName",
      defaultValue: initialValues.serverSettingName || null,
      description: {
        type: "text",
        required: true,
        read_only: false,
        label: "Server Setting Name",
        help_text:
          "The label that will be used for this server connection setting."
      }
    },
    {
      name: "serverID",
      defaultValue: initialValues.serverID || null,
      description: {
        type: "hidden",
        required: false,
        read_only: false,
        label: "Server ID",
        help_text: "Hidden Server ID"
      }
    },
    {
      name: "hostname",
      defaultValue: initialValues.hostname || null,
      description: {
        type: "text",
        required: true,
        read_only: false,
        label: "Server Hostname",
        help_text:
          "A hostname or IP address, example localhost, serial-box.com, or 192.168.5.10."
      }
    },
    {
      name: "port",
      defaultValue: initialValues.port || null,
      description: {
        type: "number",
        required: true,
        read_only: false,
        label: "Port Number",
        min_value: 1,
        max_value: 65535,
        help_text: "A port to connect to. Example, 80, 8080, 443, ..."
      }
    },
    {
      name: "path",
      defaultValue: initialValues.path || "",
      description: {
        type: "text",
        required: false,
        read_only: false,
        label: "Root Path",
        helperText: "A path to interact with API (Optional), example /api"
      }
    },
    {
      name: "ssl",
      defaultValue: initialValues.ssl || false,
      description: {
        type: "boolean",
        required: false,
        read_only: false,
        label: "SSL/TLS",
        helperText: "SSL/TLS encryption"
      }
    },
    {
      name: "username",
      defaultValue: initialValues.username || null,
      description: {
        type: "text",
        required: false,
        read_only: false,
        label: "Username",
        help_text: "Basic Auth Username"
      }
    },
    {
      name: "password",
      defaultValue: initialValues.password || null,
      description: {
        type: "password",
        required: false,
        read_only: false,
        label: "Password",
        help_text: "Basic Auth Password"
      }
    }
  ];
  getArrayFields = () => {
    return [
      {name: "protocol", value: this.protocol, editable: true},
      {name: "port", value: this.port, editable: true},
      {name: "path", value: this.path, editable: true},
      {name: "ssl", value: this.ssl, editable: true},
      {name: "url", value: this.url, editable: false},
      {name: "hostname", value: this.hostname, editable: true},
      {
        name: "serverSettingsName",
        value: this.serverSettingName,
        editable: true
      },
      {name: "username", value: this.username},
      {name: "serverID", value: this.serverID}
    ];
  };

  toJSON() {
    return {
      serverID: this.serverID,
      protocol: this.protocol,
      port: this.port,
      path: this.path,
      ssl: this.ssl,
      hostname: this.hostname,
      serverSettingName: this.serverSettingName,
      url: this.url,
      appList: this.appList,
      username: this.username,
      password: this.password
    };
  }

  getServerURL = () => {
    return `${this.protocol}://${this.hostname}:${this.port}/${this.path}`;
  };

  parseSchema = () => {
    return new Promise((resolve, reject) => {
      Swagger(this.url + "schema/")
        .then(client => {
          // swagger-js client is available.
          // client.spec / client.originalSpec / client.errors
          resolve(client);
        })
        .catch(error => {
          reject(error);
        });
    });
  };

  listApps = () => {
    if (this.appList.length === 0) {
      this.getClient()
        .then(client => {
          this.appList = Object.keys(client.apis);
          // let redux know we got our data
          store.dispatch({
            type: actions.appsListUpdated,
            payload: this.toJSON()
          });
        })
        .catch(error => {
          // handle legacy.
          this.appList = ["serialbox"];
          // let redux know we got our data
          store.dispatch({
            type: actions.appsListUpdated,
            payload: this.toJSON()
          });
        });
    }
  };

  getClient = () => {
    return new Promise((resolve, reject) => {
      if (this._client) {
        resolve(this._client);
      } else {
        // we don't have a client yet.
        // Fetch it.
        this.parseSchema()
          .then(client => {
            this._client = client;
            resolve(this._client);
          })
          .catch(error => {
            reject(error);
          });
      }
    });
  };

  getManifest = () => {
    new Promise((resolve, reject) => {
      if (!this.manifest) {
        this.getClient().then(client => {
          client.apis.manifest.quartet_manifest_list().then(result => {
            this.manifest = result.body;
            resolve(result.body);
          });
        });
      } else {
        resolve(this.manifest);
      }
    });
  };
}
