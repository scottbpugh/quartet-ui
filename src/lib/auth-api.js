// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {showMessage} from "lib/message";
import {getSyncValidators} from "components/elements/forms";
import {pluginRegistry} from "plugins/pluginRegistration";

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers.
 */
export const prepHeaders = (server, method = "GET") => {
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers and basic auth.
 */
export const prepHeadersAuth = async (server, method = "GET") => {
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * getFormInfo - Description
 *
 * @param {A server object} server       Description
 * @param {Path to endpoint} path         Description
 * @param {Callback to set the state with formStructure} createForm   Description
 * @param {type} processField Description
 *
 */
export const getFormInfo = async (server, path, createForm, processField) => {
  return fetch(`${server.url}${path}`, await prepHeadersAuth(server, "OPTIONS"))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      // parse the values and filter to the one that are not readonly.
      let postFields = data.actions.POST;
      let formStructure = Object.keys(postFields)
        .map(field => {
          if (postFields[field].read_only === false) {
            return {name: field, description: postFields[field]};
          } else {
            return null;
          }
        })
        .filter(fieldObj => {
          if (processField) {
            return processField(fieldObj);
          }
          if (fieldObj) {
            // create sync validation arrays.
            fieldObj.validate = getSyncValidators(fieldObj);
            return true;
          }
          return false;
        });
      createForm(formStructure);
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "app.servers.errorFormFetch",
        values: {error: error, serverName: server.serverSettingName}
      });
      throw error;
    });
};
