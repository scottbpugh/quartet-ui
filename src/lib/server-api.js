// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {showMessage} from "lib/message";
import {getSyncValidators} from "components/elements/forms";
import {pluginRegistry} from "plugins/pluginRegistration";

/**
 * prepHeadersAuth - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers and basic auth.
 */
export const prepHeadersAuth = async (server, method = "GET") => {
  const headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  return {
    method,
    headers,
    mode: "cors"
  };
};

/**
 * getFormInfo - Returns an object with a form structure to be used by PageForm/redux form
 *
 * @param {A server object} server       Description
 * @param {Path to endpoint} path         Description
 * @param {Callback to set the state with formStructure} createForm   Description
 * @param {type} processField Description
 *
 */
export const getFormInfo = async (server, path, createForm, processField) => {
  return fetch(`${server.url}${path}`, await prepHeadersAuth(server, "OPTIONS"))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      // parse the values and filter to the one that are not readonly.
      try {
        const postFields = data.actions.POST;
        const formStructure = Object.keys(postFields)
          .map(field => {
            if (
              field === "id" ||
              (postFields[field].type === "field" &&
                postFields[field].read_only === true)
            ) {
              return null;
            }
            return {name: field, description: postFields[field]};
          })
          .filter(fieldObj => {
            if (processField) {
              return processField(fieldObj);
            }
            if (fieldObj) {
              // create sync validation arrays.
              fieldObj.validate = getSyncValidators(fieldObj);
              return true;
            }
            return false;
          });
        createForm(formStructure);
      } catch (e) {
        if (e.message.startsWith("Cannot read property 'POST'")) {
          pluginRegistry.getHistory().push("/access-denied");
        }
      }
    })
    .catch(error => {
      debugger;
      console.log("About to redirect to access denied");
      //pluginRegistry.getHistory().push("/access-denied");
      showMessage({
        type: "danger",
        id: "app.servers.errorFormFetch",
        values: {error, serverName: server.serverSettingName}
      });
      throw error;
    });
};

/**
 * fetchObject - fetches a single object from any Swagger API endpoint.
 *
 * @param {Server}   serverInstance  A Server instance
 * @param {string} [operationId=]  The Swagger operation id
 * @param {object} [parameters={}] Parameter passed with the call.
 *
 * @return {object} An object from the API call.
 */
export const fetchObject = async (
  serverInstance,
  operationId = "",
  parameters = {}
) => {
  const client = await serverInstance.getClient();
  try {
    const response = await client.execute({
      operationId,
      parameters,
      securities: {
        authorized: client.securities,
        specSecurity: [client.spec.securityDefinitions]
      }
    });
    if (response.ok) {
      return response.body;
    }
  } catch (e) {
    showMessage({
      type: "error",
      id: "plugins.numberRange.errorVanilla",
      values: {error: e}
    });
    throw e;
  }
};

/**
 * fetchPageList - Fetches a single list page.
 *
 * @param {type} serverInstance A Server instance
 * @param {type} operationId    The Swagger
 * @param {type} parameters     Parameters passed to the call (page...)
 *
 * @return {object} A response body
 */
export const fetchPageList = async (
  serverInstance,
  operationId,
  parameters
) => {
  const client = await serverInstance.getClient();
  try {
    const response = await client.execute({
      operationId,
      parameters,
      securities: {
        authorized: client.securities,
        specSecurity: [client.spec.securityDefinitions]
      }
    });
    if (response.ok) {
      return response.body;
    }
    throw new Error(response);
  } catch (e) {
    showMessage({
      type: "error",
      id: "plugins.numberRange.errorVanilla",
      values: {error: e}
    });
    throw e;
  }
};

/**
 * fetchListAll - Loops through all pages of a list and concats the result.
 *
 * @param {Server}   serverInstance  A Server Instance
 * @param {string} [operationId=]  Swagger operation id
 * @param {object} [parameters={}] Parameters passed to the call.
 * @param {array}  [results=[]]    The results, all objects for this API list.
 *
 * @return {type} Description
 */
export const fetchListAll = async (
  serverInstance,
  operationId = "",
  parameters = {},
  results = []
) => {
  return new Promise((resolve, reject) => {
    serverInstance
      .getClient()
      .then(client => {
        client
          .execute({
            operationId,
            parameters,
            securities: {
              authorized: client.securities,
              specSecurity: [client.spec.securityDefinitions]
            }
          })
          .then(response => {
            if (response.ok) {
              if (Array.isArray(response.body)) {
                results = results.concat(response.body);
                resolve(results);
              } else if (response.body.results) {
                results = results.concat(response.body.results);
                // pagination in effect. Assuming page number navigation.
                if (response.body.next) {
                  const url = new URL(response.body.next);
                  const page = new URLSearchParams(url.search).get("page");
                  const subParameters = {...parameters, page};
                  fetchListAll(operationId, subParameters, results)
                    .then(resolve)
                    .catch(reject);
                } else {
                  resolve(results);
                }
              }
            } else {
              reject(response);
            }
          })
          .catch(e => {
            reject(e);
          });
      })
      .catch(e => {
        reject(e);
      });
  });
};

/**
 * deleteObject - Deletes an object.
 *
 * @param {string} [operationId=]  The operationId to execute.
 * @param {object} [parameters={}] Object data
 *
 * @return {object} response.
 */
export const deleteObject = async (
  serverInstance,
  operationId = "",
  parameters = {}
) => {
  const client = await serverInstance.getClient();
  const response = await client.execute({
    operationId,
    parameters,
    securities: {
      authorized: client.securities,
      specSecurity: [client.spec.securityDefinitions]
    }
  });
  if (response.ok) {
    return response;
  }
};

window.qu4rtet.exports("lib/server-api", this);
