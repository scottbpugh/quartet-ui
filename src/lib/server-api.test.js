// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import "tools/mockStore"; // mock ipcRenderer, localStorage, ...
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {MemoryRouter as Router, withRouter} from "react-router-dom";
import {IntlProvider, intlReducer} from "react-intl-redux";
import {Provider} from "react-redux";
import {mockStore, TestWrapper, initialState} from "tools/mockStore";
import {pluginRegistry} from "plugins/pluginRegistration";
import {Server} from "lib/servers";
import sinon from "sinon";
import {getFormInfo} from "lib/server-api";

// turned off since we hide the headers now.
/*it("yields the correct header object with auth", () => {
  const store = mockStore(initialState);
  const props = {
    match: {
      params: {
        serverID: "d0246781-67c6-474b-8ab0-29de61b6e6bb"
      }
    }
  };
  pluginRegistry.registerServer(
    new Server(
      initialState.serversettings.servers[
        "d0246781-67c6-474b-8ab0-29de61b6e6bb"
      ]
    )
  );
  const s = pluginRegistry.getServer("d0246781-67c6-474b-8ab0-29de61b6e6bb");
  sinon.stub(s, "getAuthorization").callsFake(() => {
    return `Token someToken`;
  });
  prepHeadersAuth(s, "GET").then(object => {
    expect(object).toMatchSnapshot();
    expect(JSON.stringify(object)).toMatchSnapshot();
  });
});*/

it("getFormInfo returns a properly structured object from a django options call (mocked)", async () => {
  const store = mockStore(initialState);
  const props = {
    match: {
      params: {
        serverID: "d0246781-67c6-474b-8ab0-29de61b6e6bb"
      }
    }
  };
  pluginRegistry.registerServer(
    new Server(
      initialState.serversettings.servers[
        "d0246781-67c6-474b-8ab0-29de61b6e6bb"
      ]
    )
  );
  const s = pluginRegistry.getServer("d0246781-67c6-474b-8ab0-29de61b6e6bb");
  sinon.stub(s, "getAuthorization").callsFake(() => {
    return `Token someToken`;
  });

  const stubbedFetch = sinon.stub(window, "fetch").callsFake(() => {
    return Promise.resolve({
      json: () =>
        Promise.resolve({
          name: "Company List",
          description: "CRUD ready model view for the Company model.",
          renders: ["application/json", "text/html", "application/xml"],
          parses: [
            "application/json",
            "application/x-www-form-urlencoded",
            "multipart/form-data"
          ],
          actions: {
            POST: {
              id: {
                type: "integer",
                required: false,
                read_only: true,
                label: "ID"
              },
              GLN13: {
                type: "string",
                required: false,
                read_only: false,
                label: "GLN13",
                help_text:
                  "The GLN (Global Location Number) provides a standard means to identify legal entities, trading parties and locations to support the requirements of electronic commerce. The GLN-13 is defined by GS1",
                max_length: 13
              },
              SGLN: {
                type: "string",
                required: false,
                read_only: false,
                label: "SGLN",
                help_text:
                  "The SGLN EPC scheme is used to assign a unique identity to a physical location or sub-location, such as a specific building or a specific unit of shelving within a warehouse.  TheSGLN is expressed as a URN value.",
                max_length: 150
              },
              name: {
                type: "string",
                required: true,
                read_only: false,
                label: "Name",
                help_text: "A unique name for the location or party.",
                max_length: 128
              },
              address1: {
                type: "string",
                required: false,
                read_only: false,
                label: "Street Address One",
                help_text:
                  "For example, the name of the street and the number in the street or the name of a building",
                max_length: 1000
              },
              address2: {
                type: "string",
                required: false,
                read_only: false,
                label: "Street Address Two",
                help_text:
                  "The second free form line complements the first free form line to locate the party or location.",
                max_length: 1000
              },
              address3: {
                type: "string",
                required: false,
                read_only: false,
                label: "Street Address Three",
                help_text:
                  "The third free form line complements the first and second free form lines where necessary.",
                max_length: 1000
              },
              country: {
                type: "string",
                required: false,
                read_only: false,
                label: "Country Code",
                help_text: "Country ISO 3166-1 alpha-2 Code",
                max_length: 2
              },
              city: {
                type: "string",
                required: false,
                read_only: false,
                label: "City",
                help_text: "City",
                max_length: 50
              },
              state_province: {
                type: "string",
                required: false,
                read_only: false,
                label: "State or Province",
                help_text:
                  "One of the constituent units of a nation having a federal government.",
                max_length: 20
              },
              postal_code: {
                type: "string",
                required: false,
                read_only: false,
                label: "Postal Code",
                help_text: "Postal Code",
                max_length: 20
              },
              latitude: {
                type: "decimal",
                required: false,
                read_only: false,
                label: "Latitude",
                help_text:
                  " Latitude of the location, in degrees. Positive numbers are northern latitude; negative numbers are southern latitude."
              },
              longitude: {
                type: "decimal",
                required: false,
                read_only: false,
                label: "Longitude",
                help_text:
                  "Longitude of the location, in degrees. Positive numbers are eastern longitude; negative numbers are western longitude."
              },
              gs1_company_prefix: {
                type: "string",
                required: true,
                read_only: false,
                label: "GS1 Company Prefix",
                help_text:
                  "A GS1 Company Prefix is a unique string of four to twelve digits used to issue GS1 identification keys.",
                max_length: 12
              },
              company_type: {
                type: "field",
                required: false,
                read_only: false,
                label: "Type",
                help_text: "Describes the type of company."
              }
            }
          }
        })
    });
  });
  await getFormInfo(s, "/some-path/", formStructure => {
    expect(formStructure).toMatchSnapshot();
  });
});
