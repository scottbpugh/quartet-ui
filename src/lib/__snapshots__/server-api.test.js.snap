// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFormInfo returns a properly structured object from a django options call (mocked) 1`] = `
Array [
  Object {
    "description": Object {
      "help_text": "The GLN (Global Location Number) provides a standard means to identify legal entities, trading parties and locations to support the requirements of electronic commerce. The GLN-13 is defined by GS1",
      "label": "GLN13",
      "max_length": 13,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "GLN13",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "The SGLN EPC scheme is used to assign a unique identity to a physical location or sub-location, such as a specific building or a specific unit of shelving within a warehouse.  TheSGLN is expressed as a URN value.",
      "label": "SGLN",
      "max_length": 150,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "SGLN",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "A unique name for the location or party.",
      "label": "Name",
      "max_length": 128,
      "read_only": false,
      "required": true,
      "type": "string",
    },
    "name": "name",
    "validate": Array [
      [Function],
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "For example, the name of the street and the number in the street or the name of a building",
      "label": "Street Address One",
      "max_length": 1000,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "address1",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "The second free form line complements the first free form line to locate the party or location.",
      "label": "Street Address Two",
      "max_length": 1000,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "address2",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "The third free form line complements the first and second free form lines where necessary.",
      "label": "Street Address Three",
      "max_length": 1000,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "address3",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "Country ISO 3166-1 alpha-2 Code",
      "label": "Country Code",
      "max_length": 2,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "country",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "City",
      "label": "City",
      "max_length": 50,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "city",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "One of the constituent units of a nation having a federal government.",
      "label": "State or Province",
      "max_length": 20,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "state_province",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "Postal Code",
      "label": "Postal Code",
      "max_length": 20,
      "read_only": false,
      "required": false,
      "type": "string",
    },
    "name": "postal_code",
    "validate": Array [
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": " Latitude of the location, in degrees. Positive numbers are northern latitude; negative numbers are southern latitude.",
      "label": "Latitude",
      "read_only": false,
      "required": false,
      "type": "decimal",
    },
    "name": "latitude",
    "validate": Array [],
  },
  Object {
    "description": Object {
      "help_text": "Longitude of the location, in degrees. Positive numbers are eastern longitude; negative numbers are western longitude.",
      "label": "Longitude",
      "read_only": false,
      "required": false,
      "type": "decimal",
    },
    "name": "longitude",
    "validate": Array [],
  },
  Object {
    "description": Object {
      "help_text": "A GS1 Company Prefix is a unique string of four to twelve digits used to issue GS1 identification keys.",
      "label": "GS1 Company Prefix",
      "max_length": 12,
      "read_only": false,
      "required": true,
      "type": "string",
    },
    "name": "gs1_company_prefix",
    "validate": Array [
      [Function],
      [Function],
    ],
  },
  Object {
    "description": Object {
      "help_text": "Describes the type of company.",
      "label": "Type",
      "read_only": false,
      "required": false,
      "type": "field",
    },
    "name": "company_type",
    "validate": Array [],
  },
]
`;

exports[`yields the correct header object with auth 1`] = `
Object {
  "headers": Headers {
    "_headers": Object {
      "accept": Array [
        "application/json",
      ],
      "authorization": Array [
        "Token someToken",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
  "method": "GET",
  "mode": "cors",
}
`;

exports[`yields the correct header object with auth 2`] = `"{\\"method\\":\\"GET\\",\\"headers\\":{\\"_headers\\":{\\"accept\\":[\\"application/json\\"],\\"content-type\\":[\\"application/json\\"],\\"authorization\\":[\\"Token someToken\\"]}},\\"mode\\":\\"cors\\"}"`;
